src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -454,7 +454,7 @@ async def create_tool_runner(
 
     append_only_caches = {
         **merged_extras.append_only_caches,
-        **(request.named_caches or {}),
+        **(request.named_caches or {}),  # type: ignore[dict-item]
```

Comment:
It looks like it's this `or` construct that's causing the issues, likely because the `{}` becomes `dict[Never, Never]` and that apparently doesn't satisfy `SupportsKeysAndGetItem[K, V]` with the specific `K` and `V` types.

```python
def direct_splat_frozen_only(fd: FrozenDict[str, int]) -> dict[str, int]:
    return {**fd} # okay

def direct_splat_dict_too(fd: FrozenDict[str, int] | dict[str, int]) -> dict[str, int]:
    return {**fd} # okay
    
def or_splat_explicit(fd: FrozenDict[str, int]) -> dict[str, int]:
    empty: dict[str, int] = {}
    return {**(fd or empty)} # okay

def or_splat_implicit(fd: None | FrozenDict[str, int]) -> dict[str, int]:
    return {**(fd or {})} # error: Unpacked dict entry 0 has incompatible type "FrozenDict[str, int] | dict[Never, Never]"; expected "SupportsKeysAndGetItem[str, int]"  [dict-item]
```

https://mypy-play.net/?mypy=latest&python=3.12&gist=661f39bc47fc9b1b4f8a80fe0814d1c1

I filed https://github.com/python/mypy/issues/17790

Happy to go with what you've got for now.

Reply:
Interesting. Yeah, let's ignore it for now. If and when it gets fixed upstream, we should get a "unused type ignore" error once the ignore is no longer necessary.